retryablehttp/retryablehttp.go

```diff
@@ -0,0 +1,149 @@
+		InitialInterval: conf.GetReloadableDurationVar(100, time.Millisecond, "retryablehttp.initialInterval").Load(),
+		MaxInterval:     conf.GetReloadableDurationVar(1000, time.Millisecond, "retryablehttp.maxInterval").Load(),
+		MaxElapsedTime:  conf.GetReloadableDurationVar(10, time.Second, "retryablehttp.maxElapsedTime").Load(),
+		Multiplier:      conf.GetReloadableFloat64Var(1.5, "retryablehttp.multiplier").Load(),
+	}
+}
+
+type retryableHTTPClient struct {
+	requestDoer
+	config *Config
+	// onFailure is called when a retryable error occurs
+	onFailure func(err error, duration time.Duration)
+}
+
+type Option func(*retryableHTTPClient)
+
+func WithRequestDoer(client requestDoer) Option {
+	return func(retryableHTTPClient *retryableHTTPClient) {
+		retryableHTTPClient.requestDoer = client
+	}
+}
+
+func WithOnFailure(onFailure func(err error, duration time.Duration)) Option {
+	return func(retryableHTTPClient *retryableHTTPClient) {
+		retryableHTTPClient.onFailure = onFailure
+	}
+}
+
+// NewRetryableHTTPClient creates a new retryable HTTP client with the specified configuration and options.
+// It uses the `backoff` package to implement retry logic for HTTP requests.
+// All the 5xx and 429 errors will be retried
+// The default retry strategy is exponential backoff
+// Parameters:
+// - config: Configuration for the exponential backoff strategy
+// - options: Optional functional options to further configure the client
+//
+// Returns:
+// - HttpClient: A configured retryable HTTP client
+//
+// The client is initialized with a transport that has:
+// - Keep-alives disabled
+// - Maximum 100 connections per host
+// - Maximum 10 idle connections per host
+// - Idle connection timeout of 30 seconds
+func NewRetryableHTTPClient(config *Config, options ...Option) HttpClient {
+	if config == nil {
+		config = NewDefaultConfig()
+	}
+	httpClient := &retryableHTTPClient{
+		requestDoer: &http.Client{
+			Transport: &http.Transport{
+				DisableKeepAlives:   true,
+				MaxConnsPerHost:     100,
+				MaxIdleConnsPerHost: 10,
+				IdleConnTimeout:     30 * time.Second,
+			},
+		},
+		config: config,
+	}
+	for _, option := range options {
+		option(httpClient)
+	}
+	return httpClient
+}
+
+// Do executes an HTTP request with retry logic.
+func (c *retryableHTTPClient) Do(method, url string, body io.Reader, headers map[string]string) (*http.Response, error) {
+	var (
+		resp *http.Response
+		err  error
+	)
+
+	_ = backoff.RetryNotify(
+		func() error {
+			var req *http.Request
+			req, err = http.NewRequest(method, url, body)
+			if err == nil {
+				for key, value := range headers {
+					req.Header.Set(key, value)
+				}
+				resp, err = c.requestDoer.Do(req) // nolint: bodyclose
+				// retry 5xx errors
+				if err == nil && (resp.StatusCode >= http.StatusInternalServerError || resp.StatusCode == http.StatusTooManyRequests) {
+					return fmt.Errorf("non-success status code: %d", resp.StatusCode)
+				}
```

Comment:
this behaviour should be:
- well documented in the client's docs
- extensible

Reply:
we can also take some inspiration from [here](https://github.com/hashicorp/go-retryablehttp/blob/main/client.go#L495-L542)